controllers/configuration/backend/backend.go

```diff
@@ -0,0 +1,335 @@
+		if v == backendType {
+			return true
+		}
+	}
+	return false
+}
+
+// ParseConfigurationBackend returns backend hcl string, backend type, useCustom, secretRef list and error
+func ParseConfigurationBackend(configuration *v1beta2.Configuration, terraformBackendNamespace string) (*Conf, error) {
+	backend := configuration.Spec.Backend
+
+	switch {
+
+	case backend != nil && len(backend.Inline) > 0:
+		// In this case, use the inline custom backend
+		backendTF, backendType, err := handleInlineBackendHCL(backend.Inline)
+		if err != nil {
+			return nil, err
+		}
+		return &Conf{
+			BackendType: backendType,
+			HCL:         backendTF,
+			UseCustom:   true,
+		}, nil
+
+	case backend != nil && len(backend.BackendType) > 0:
+		// In this case, use the explicit custom backend
+
+		// first, check if is valid custom backend
+
+		backendType := strings.ToLower(backend.BackendType)
+		// check if backendType is valid
+		if !checkBackendTypeValid(backendType) {
+			return nil, fmt.Errorf("%s is unsupported backendType", backend.BackendType)
+		}
+		// fetch backendConfValue using reflection
+		backendStructValue := reflect.ValueOf(backend)
+		if backendStructValue.Kind() == reflect.Ptr {
+			backendStructValue = backendStructValue.Elem()
+		}
+		backendField := backendStructValue.FieldByNameFunc(func(name string) bool {
+			return strings.EqualFold(name, backendType)
+		})
+		if backendField.IsNil() {
+			return nil, fmt.Errorf("there is no configuration for backendType %s", backend.BackendType)
+		}
+		backendConfValue := backendField.Interface()
+
+		// second, handle the backendConf
+		backendHCL, backendType, secretList := handleExplicitBackend(backendConfValue, backendType)
+		return &Conf{
+			BackendType: backendType,
+			HCL:         backendHCL,
+			UseCustom:   true,
+			Secrets:     secretList,
+		}, nil
+
+	case backend != nil && len(backend.BackendType) == 0:
+		// In this case, use the default k8s backend
+		klog.Warningf("the spec.backend.backend_type of Configuration{Namespace: %s, Name: %s} is empty, use the default kubernetes backend", configuration.Namespace, configuration.Name)
+		fallthrough // down to default
+
+	default:
+		// use the default k8s backend
+		if configuration.Spec.Backend != nil {
+			if configuration.Spec.Backend.SecretSuffix == "" {
+				configuration.Spec.Backend.SecretSuffix = configuration.Name
+			}
+			configuration.Spec.Backend.InClusterConfig = true
+		} else {
+			configuration.Spec.Backend = &v1beta2.Backend{
+				SecretSuffix:    configuration.Name,
+				InClusterConfig: true,
+			}
+		}
+		backendTF, err := renderTemplate(configuration.Spec.Backend, terraformBackendNamespace)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to prepare Terraform backend configuration")
+		}
+		return &Conf{
+			BackendType: "kubernetes",
+			HCL:         backendTF,
+			UseCustom:   false,
+		}, nil
+	}
+
+}
+
+func handleInlineBackendHCL(code string) (string, string, error) {
```

Comment:
Why do we need to handle inlined backend HCL? How about transparently storing it in backend.tf?

Reply:
This function is used to check for illegal properties in the inline hcl code.